File: ChatClientCS/ViewModels/MainWindowViewModel.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using System.Collections.ObjectModel;
using System.IO;
using System.Drawing;
using ChatClientCS.Services;
using ChatClientCS.Enums;
using ChatClientCS.Models;
using ChatClientCS.Commands;
using System.Windows.Input;
using System.Windows.Media;
using System.Diagnostics;
using System.Reactive.Linq;

namespace ChatClientCS.ViewModels
{
    public class MainWindowViewModel : ViewModelBase
    {
        private IChatService chatService;
        private IDialogService dialogService;
        private TaskFactory ctxTaskFactory;
        private const int MAX_IMAGE_WIDTH = 150;
        private const int MAX_IMAGE_HEIGHT = 150;

        private string _nomeUtente;
        public string NomeUtente
        {
            get { return _nomeUtente; }
            set
            {
                _nomeUtente = value;
                OnPropertyChanged();
            }
        }

        private string _imgProfilo;
        public string imgProfilo
        {
            get { return _imgProfilo; }
            set
            {
                _imgProfilo = value;
                OnPropertyChanged();
            }
        }

        private ObservableCollection<Partecipante> _partecipanti = new ObservableCollection<Partecipante>();
        public ObservableCollection<Partecipante> Partecipanti
        {
            get { return _partecipanti; }
            set
            {
                _partecipanti = value;
                OnPropertyChanged();
            }
        }

        private Partecipante _partecipanteSelezionato;
        public Partecipante PartecipanteSelezionato
        {
            get { return _partecipanteSelezionato; }
            set
            {
                _partecipanteSelezionato = value;
                if (PartecipanteSelezionato.HaInviatoNuovoMessaggio) PartecipanteSelezionato.HaInviatoNuovoMessaggio = false;
                OnPropertyChanged();
            }
        }

        private ModalitaUser _modalitaUser;
        public ModalitaUser ModalitaUser
        {
            get { return _modalitaUser; }
            set
            {
                _modalitaUser = value;
                OnPropertyChanged();
            }
        }

        private string _msgTesto;
        public string MsgTesto
        {
            get { return _msgTesto; }
            set
            {
                _msgTesto = value;
                OnPropertyChanged();
            }
        }

        private bool _connesso;
        public bool Connesso
        {
            get { return _connesso; }
            set
            {
                _connesso = value;
                OnPropertyChanged();
            }
        }

        private bool _loggato;
        public bool Loggato
        {
            get { return _loggato; }
            set
            {
                _loggato = value;
                OnPropertyChanged();
            }
        }

        #region Operazioni di connessione
        private ICommand _opConnessione;
        public ICommand OpConnessione
        {
            get
            {
                return _opConnessione ?? (_opConnessione = new RelayCommandAsync(() => Connetti()));
            }
        }

        private async Task<bool> Connetti()
        {
            try
            {
                await chatService.ConnessioneAsync();
                Connesso = true;
                return true;
            }
            catch (Exception) { return false; }
        }
        #endregion

        #region Comandi di accesso
        private ICommand _comandoLogin;
        public ICommand ComandoAccesso
        {
            get
            {
                return _comandoLogin ?? (_comandoLogin =
                    new RelayCommandAsync(() => Accesso(), (o) => PuoAccedere()));
            }
        }

        private async Task<bool> Accesso()
        {
            try
            {
                List<Utente> utenti = new List<Utente>();
                utenti = await chatService.AccessoAsincr(_nomeUtente, Avatar());
                if (utenti != null)
                {
                    utenti.ForEach(u => Partecipanti.Add(new Partecipante { Nome = u.Nome, imgProfilo = u.imgProfilo }));
                    ModalitaUser = ModalitaUser.Chat;
                    Loggato = true;
                    return true;
                }
                else
                {
                    dialogService.ShowNotification("Nome utente già in uso");
                    return false;
                }

            }
            catch (Exception) { return false; }
        }

        private bool PuoAccedere()
        {
            return !string.IsNullOrEmpty(NomeUtente) && NomeUtente.Length >= 2 && Connesso;
        }
        #endregion

        #region Comandi di Logout
        private ICommand _opLogout;
        public ICommand OpLogout
        {
            get
            {
                return _opLogout ?? (_opLogout =
                    new RelayCommandAsync(() => Logout(), (o) => CanLogout()));
            }
        }

        private async Task<bool> Logout()
        {
            try
            {
                await chatService.DisconnessioneAsincr();
                ModalitaUser = ModalitaUser.Login;
                return true;
            }
            catch (Exception) { return false; }
        }

        private bool CanLogout()
        {
            return Connesso && Loggato;
        }
        #endregion

        #region Comandi Scrittura
        private ICommand _opStaScivendo;
        public ICommand OpStaScrivendo
        {
            get
            {
                return _opStaScivendo ?? (_opStaScivendo =
                    new RelayCommandAsync(() => StaScrivendo(), (o) => PuoUsareStaScrivendo()));
            }
        }

        private async Task<bool> StaScrivendo()
        {
            try
            {
                await chatService.StaScrivendoAsincr(PartecipanteSelezionato.Nome);
                return true;
            }
            catch (Exception) { return false; }
        }

        private bool PuoUsareStaScrivendo()
        {
            return (PartecipanteSelezionato != null && PartecipanteSelezionato.Loggato);
        }
        #endregion

        #region Comando per inviare messaggi di testo
        private ICommand _opInviaMsgTesto;
        public ICommand OpInviaMsgTesto
        {
            get
            {
                return _opInviaMsgTesto ?? (_opInviaMsgTesto =
                    new RelayCommandAsync(() => InviaMsgTesto(), (o) => PuoInviareMsgTesto()));
            }
        }

        private async Task<bool> InviaMsgTesto()
        {
            try
            {
                var destinatario = _partecipanteSelezionato.Nome;
                await chatService.MandaMsgUnicastAsincr(destinatario, _msgTesto);
                return true;
            }
            catch (Exception) { return false; }
            finally
            {
                MessaggioChat msg = new MessaggioChat
                {
                    Autore = NomeUtente,
                    Testo = _msgTesto,
                    DataOra = DateTime.Now,
                    IsOriginNative = true
                };
                PartecipanteSelezionato.Chatter.Add(msg);
                MsgTesto = string.Empty;
            }
        }

        private bool PuoInviareMsgTesto()
        {
            return (!string.IsNullOrEmpty(MsgTesto) && Connesso &&
                _partecipanteSelezionato != null && _partecipanteSelezionato.Loggato);
        }
        #endregion

        #region Comando per inviare immagini
        private ICommand _opInviaMsgImmagine;
        public ICommand OpInviaMsgImmagine
        {
            get
            {
                return _opInviaMsgImmagine ?? (_opInviaMsgImmagine =
                    new RelayCommandAsync(() => InviaMsgImmagine(), (o) => PuoInviareMsgImmagine()));
            }
        }

        private async Task<bool> InviaMsgImmagine()
        {
            var immagine = dialogService.OpenFile("Seleziona file immagine", "Immagini (*.jpg;*.png)|*.jpg;*.png");
            if (string.IsNullOrEmpty(immagine)) return false;

            var img = await Task.Run(() => File.ReadAllBytes(immagine));

            try
            {
                var destinatario = _partecipanteSelezionato.Nome;
                await chatService.MandaMsgUnicastAsincr(destinatario, img);
                return true;
            }
            catch (Exception) { return false; }
            finally
            {
                MessaggioChat msg = new MessaggioChat { Autore = NomeUtente, Immagine = immagine, DataOra = DateTime.Now, IsOriginNative = true };
                PartecipanteSelezionato.Chatter.Add(msg);
            }           
        }

        private bool PuoInviareMsgImmagine()
        {
            return (Connesso && _partecipanteSelezionato != null && _partecipanteSelezionato.Loggato);
        }
        #endregion

        #region Operazione Selezione Immagine Profilo
        private ICommand _opSelezionaImgProfilo;
        public ICommand OpSelezionaImgProfilo
        {
            get
            {
                return _opSelezionaImgProfilo ?? (_opSelezionaImgProfilo =
                    new RelayCommand((o) => SelezionaImgProfilo()));
            }
        }

        private void SelezionaImgProfilo()
        {
            var _imgProfilo = dialogService.OpenFile("Seleziona file immagine", "Immagini (*.jpg;*.png)|*.jpg;*.png");
            if (!string.IsNullOrEmpty(_imgProfilo))
            {
                var img = Image.FromFile(_imgProfilo);
                if (img.Width > MAX_IMAGE_WIDTH || img.Height > MAX_IMAGE_HEIGHT)
                {
                    dialogService.ShowNotification($"Le dimensioni della immagine devono esesere di {MAX_IMAGE_WIDTH} x {MAX_IMAGE_HEIGHT} o inferiori.");
                    return;
                }
                imgProfilo = _imgProfilo;
            }
        }
        #endregion

        #region Comando per aprire immagini
        private ICommand _opApriImmagine;
        public ICommand OpApriImmagine
        {
            get
            {
                return _opApriImmagine ?? (_opApriImmagine =
                    new RelayCommand<MessaggioChat>((m) => ApriImmagine(m)));
            }
        }

        private void ApriImmagine(MessaggioChat messaggio)
        {
            var img = messaggio.Immagine;
            if (string.IsNullOrEmpty(img) || !File.Exists(img)) return;
            Process.Start(img);
        }
        #endregion

        #region Gestore Eventi (Event Handlers)
        private void NuovoMsgTesto(string nome, string testo, TipoDiMessaggio tipoMsg)
        {
            if (tipoMsg == TipoDiMessaggio.Unicast)
            {
                MessaggioChat msgChat = new MessaggioChat { Autore = nome, Testo = testo, DataOra = DateTime.Now };
                var mittente = _partecipanti.Where((u) => string.Equals(u.Nome, nome)).FirstOrDefault();
                ctxTaskFactory.StartNew(() => mittente.Chatter.Add(msgChat)).Wait();

                if (!(PartecipanteSelezionato != null && mittente.Nome.Equals(PartecipanteSelezionato.Nome)))
                {
                    ctxTaskFactory.StartNew(() => mittente.HaInviatoNuovoMessaggio = true).Wait();
                }
            }
        }

        private void NuovoMsgImmagine(string nome, byte[] immagine, TipoDiMessaggio tipoMsg)
        {
            if (tipoMsg == TipoDiMessaggio.Unicast)
            {
                var CartellaImmagini = Path.Combine(Environment.CurrentDirectory, "Image Messages");
                if (!Directory.Exists(CartellaImmagini)) Directory.CreateDirectory(CartellaImmagini);

                var contImmagini = Directory.EnumerateFiles(CartellaImmagini).Count() + 1;
                var percorsoImmagine = Path.Combine(CartellaImmagini, $"IMG_{contImmagini}.jpg");

                ImageConverter convertitoreImmagini = new ImageConverter();
                using (Image img = (Image)convertitoreImmagini.ConvertFrom(immagine))
                {
                    img.Save(percorsoImmagine);
                }

                MessaggioChat msgChat = new MessaggioChat { Autore = nome, Immagine = percorsoImmagine, DataOra = DateTime.Now };
                var mittente = _partecipanti.Where(u => string.Equals(u.Nome, nome)).FirstOrDefault();
                ctxTaskFactory.StartNew(() => mittente.Chatter.Add(msgChat)).Wait();

                if (!(PartecipanteSelezionato != null && mittente.Nome.Equals(PartecipanteSelezionato.Nome)))
                {
                    ctxTaskFactory.StartNew(() => mittente.HaInviatoNuovoMessaggio = true).Wait();
                }
            }
        }

        private void LoginPartecipante(Utente u)
        {
            var _partecipante = Partecipanti.FirstOrDefault(partecipante => string.Equals(partecipante.Nome, u.Nome));
            if (_loggato && _partecipante == null)
            {
                ctxTaskFactory.StartNew(() => Partecipanti.Add(new Partecipante
                {
                    Nome = u.Nome,
                    imgProfilo = u.imgProfilo
                })).Wait();
            }
        }

        private void DisconnessionePartecipante(string nome)
        {
            var tizio = Partecipanti.Where((partecipante) => string.Equals(partecipante.Nome, nome)).FirstOrDefault();
            if (tizio != null) tizio.Loggato = false;
        }

        private void RiconnessionePartecipante(string nome)
        {
            var tizio = Partecipanti.Where((partecipante) => string.Equals(partecipante.Nome, nome)).FirstOrDefault();
            if (tizio != null) tizio.Loggato = true;
        }

        private void Riconnessione()
        {
            Connesso = false;
            Loggato = false;
        }

        private async void Riconnesso()
        {
            var imgProfilo = Avatar();
            if (!string.IsNullOrEmpty(_nomeUtente)) await chatService.AccessoAsincr(_nomeUtente, imgProfilo);
            Connesso = true;
            Loggato = true;
        }

        private async void Disconnesso()
        {
            var taskConnessione = chatService.ConnessioneAsync();
            await taskConnessione.ContinueWith(task => {
                if (!task.IsFaulted)
                {
                    Connesso = true;
                    chatService.AccessoAsincr(_nomeUtente, Avatar()).Wait();
                    Loggato = true;
                }
            });
        }

        private void PartecipanteStaScrivendo(string nome)
        {
            var tizio = Partecipanti.Where((p) => string.Equals(p.Nome, nome)).FirstOrDefault();
            if (tizio != null && !tizio.StaScrivendo)
            {
                tizio.StaScrivendo = true;
                Observable.Timer(TimeSpan.FromMilliseconds(1500)).Subscribe(t => tizio.StaScrivendo = false);
            }
        }
        #endregion

        private byte[] Avatar()
        {
            byte[] imgProfilo = null;
            if (!string.IsNullOrEmpty(_imgProfilo)) imgProfilo = File.ReadAllBytes(_imgProfilo);
            return imgProfilo;
        }

        public MainWindowViewModel(IChatService chatSvc, IDialogService diagSvc)
        {
            dialogService = diagSvc;
            chatService = chatSvc;

            chatSvc.NuovoMsgTesto += NuovoMsgTesto;
            chatSvc.NuovoMsgImmagine += NuovoMsgImmagine;
            chatSvc.PartecipanteHaFattoAccesso += LoginPartecipante;
            chatSvc.PartecipanteSiEDisconnesso += DisconnessionePartecipante;
            chatSvc.PartecipanteDisconnesso += DisconnessionePartecipante;
            chatSvc.PartecipanteRiconnesso += RiconnessionePartecipante;
            chatSvc.PartecipanteStaScrivendo += PartecipanteStaScrivendo;
            chatSvc.RiconnessioneInCorso += Riconnessione;
            chatSvc.ConnessioneRistabilita += Riconnesso;
            chatSvc.ConnessioneChiusa += Disconnesso;

            ctxTaskFactory = new TaskFactory(TaskScheduler.FromCurrentSynchronizationContext());
        }

    }
}

File: ChatClientVB/ViewModels/MainWindowViewModel.vb
Imports System.Collections.ObjectModel
Imports System.IO
Imports System.Drawing
Imports System.Reactive.Linq

Public Class MainWindowViewModel
    Inherits ViewModelBase

    Private chatService As IChatService
    Private dialogService As IDialogService
    Private ctxTaskFactory As TaskFactory
    Private Const MAX_IMAGE_WIDTH As Integer = 150
    Private Const MAX_IMAGE_HEIGHT As Integer = 150

    Private _nomeUtente As String
    Public Property NomeUtente As String
        Get
            Return _nomeUtente
        End Get
        Set(value As String)
            _nomeUtente = value
            OnPropertyChanged()
        End Set
    End Property

    Private _imgProfilo As String
    Public Property imgProfilo As String
        Get
            Return _imgProfilo
        End Get
        Set(value As String)
            _imgProfilo = value
            OnPropertyChanged()
        End Set
    End Property

    Private _partecipanti As New ObservableCollection(Of Partecipante)
    Public Property Partecipanti As ObservableCollection(Of Partecipante)
        Get
            Return _partecipanti
        End Get
        Set(value As ObservableCollection(Of Partecipante))
            _partecipanti = value
            OnPropertyChanged()
        End Set
    End Property

    Private _partecipanteSelezionato As Partecipante
    Public Property PartecipanteSelezionato As Partecipante
        Get
            Return _partecipanteSelezionato
        End Get
        Set(value As Partecipante)
            _partecipanteSelezionato = value
            If PartecipanteSelezionato.HaMandatoNuovoMsg Then PartecipanteSelezionato.HaMandatoNuovoMsg = False
            OnPropertyChanged()
        End Set
    End Property

    Private _modalitaUser As ModalitaUser
    Public Property ModalitaUser As ModalitaUser
        Get
            Return _modalitaUser
        End Get
        Set(value As ModalitaUser)
            _modalitaUser = value
            OnPropertyChanged()
        End Set
    End Property

    Private _msgTesto As String
    Public Property MsgTesto As String
        Get
            Return _msgTesto
        End Get
        Set(value As String)
            _msgTesto = value
            OnPropertyChanged()
        End Set
    End Property

    Private _connesso As Boolean
    Public Property Connesso As Boolean
        Get
            Return _connesso
        End Get
        Set(value As Boolean)
            _connesso = value
            OnPropertyChanged()
        End Set
    End Property

    Private _loggato As Boolean
    Public Property Loggato As Boolean
        Get
            Return _loggato
        End Get
        Set(value As Boolean)
            _loggato = value
            OnPropertyChanged()
        End Set
    End Property

#Region "Comandi di connessione"
    Private _opConnessione As ICommand
    Public ReadOnly Property OpConnessione As ICommand
        Get
            Return If(_opConnessione, New RelayCommandAsync(Function() Connetti()))
        End Get
    End Property

    Private Async Function Connetti() As Task(Of Boolean)
        Try
            Await chatService.ConnessioneAsync()
            Connesso = True
            Return True
        Catch eccezione As Exception
            Return False
        End Try
    End Function
#End Region

#Region "Comandi di accesso"
    Private _comandoLogin As ICommand
    Public ReadOnly Property ComandoAccesso As ICommand
        Get
            Return If(_comandoLogin, New RelayCommandAsync(Function() Accesso(), AddressOf PuoAccedere))
        End Get
    End Property

    Private Async Function Accesso() As Task(Of Boolean)
        Try
            Dim utenti As List(Of Utente)
            utenti = Await chatService.AccessoAsincr(_nomeUtente, Avatar())
            If utenti IsNot Nothing Then
                utenti.ForEach(Sub(u) Partecipanti.Add(New Partecipante With {.Nome = u.Nome, .imgProfilo = u.imgProfilo}))
                ModalitaUser = ModalitaUser.Chat
                Loggato = True
                Return True
            Else
                dialogService.ShowNotification("Nome utente già in uso")
                Return False
            End If
        Catch ex As Exception
            Return False
        End Try
    End Function

    Private Function PuoAccedere() As Boolean
        Return Not String.IsNullOrEmpty(NomeUtente) AndAlso NomeUtente.Length >= 2 AndAlso Connesso
    End Function
#End Region

#Region "Comandi di Logout"
    Private _opLogout As ICommand
    Public ReadOnly Property OpLogout As ICommand
        Get
            Return If(_opLogout, New RelayCommandAsync(Function() Logout(), AddressOf PuoDisconnettersi))
        End Get
    End Property

    Private Async Function Logout() As Task(Of Boolean)
        Try
            Await chatService.DisconnessioneAsincr()
            ModalitaUser = ModalitaUser.Accesso
            Return True
        Catch ex As Exception
            Return False
        End Try
    End Function

    Private Function PuoDisconnettersi() As Boolean
        Return Connesso AndAlso Loggato
    End Function
#End Region

#Region "Comandi Scrittura"
    Private _opStaScrivendo As ICommand
    Public ReadOnly Property OpStaScrivendo As ICommand
        Get
            Return If(_opStaScrivendo, New RelayCommandAsync(Function() StaScrivendo(), AddressOf PuoUsareStaScrivendo))
        End Get
    End Property

    Private Async Function StaScrivendo() As Task(Of Boolean)
        Try
            Await chatService.StaScrivendoAsincr(PartecipanteSelezionato.Nome)
            Return True
        Catch ex As Exception
            Return False
        End Try
    End Function

    Private Function PuoUsareStaScrivendo() As Boolean
        Return PartecipanteSelezionato IsNot Nothing AndAlso PartecipanteSelezionato.Loggato
    End Function
#End Region

#Region "Comando per inviare messaggio di testo"
    Private _opInviaMsgTesto As ICommand
    Public ReadOnly Property OpInviaMsgTesto As ICommand
        Get
            Return If(_opInviaMsgTesto, New RelayCommandAsync(Function() InviaMsgTesto(),
                                                                     AddressOf PuoInviareMsgTesto))
        End Get
    End Property

    Private Async Function InviaMsgTesto() As Task(Of Boolean)
        Try
            Dim destinatario = _partecipanteSelezionato.Nome
            Await chatService.MandaMsgUnicastAsincr(destinatario, _msgTesto)
            Return True
        Catch ex As Exception
            Return False
        Finally
            Dim msg As New MessaggioChat With {.Autore = NomeUtente, .Testo = _msgTesto,
                .DataOra = DateTime.Now, .IsOriginNative = True}
            PartecipanteSelezionato.Chatter.Add(msg)
            MsgTesto = String.Empty
        End Try
    End Function

    Private Function PuoInviareMsgTesto() As Boolean
        Return Not String.IsNullOrEmpty(MsgTesto) AndAlso Connesso AndAlso
            _partecipanteSelezionato IsNot Nothing AndAlso _partecipanteSelezionato.Loggato
    End Function
#End Region

#Region "Comando per inviare immagini"
    Private _opInviaMsgImmagine As ICommand
    Public ReadOnly Property OpInviaMsgImmagine As ICommand
        Get
            Return If(_opInviaMsgImmagine, New RelayCommandAsync(Function() InviaMsgImmagine(),
                                                                     AddressOf PuoMandareMsgImmagine))
        End Get
    End Property

    Private Async Function InviaMsgImmagine() As Task(Of Boolean)
        Dim immagine = dialogService.OpenFile("Seleziona file immagine", "Immagini (*.jpg;*.png)|*.jpg;*.png")
        If String.IsNullOrEmpty(immagine) Then Return False

        Dim img = Await Task.Run(Function() File.ReadAllBytes(immagine))

        Try
            Dim destinatario = _partecipanteSelezionato.Nome
            Await chatService.MandaMsgUnicastAsincr(destinatario, img)
            Return True
        Catch ex As Exception
            Return False
        Finally
            Dim msg As New MessaggioChat With {.Autore = NomeUtente, .Immagine = immagine, .DataOra = DateTime.Now, .IsOriginNative = True}
            PartecipanteSelezionato.Chatter.Add(msg)
        End Try
    End Function

    Private Function PuoMandareMsgImmagine() As Boolean
        Return Connesso AndAlso _partecipanteSelezionato IsNot Nothing AndAlso _partecipanteSelezionato.Loggato
    End Function
#End Region

#Region "Operazione Selezione Immagine Profilo"
    Private _opSelezionaImgProfilo As ICommand
    Public ReadOnly Property SelectPhotoCommand As ICommand
        Get
            Return If(_opSelezionaImgProfilo, New RelayCommand(AddressOf SelezionaImgProfilo))
        End Get
    End Property

    Private Sub SelezionaImgProfilo()
        Dim _imgProfilo = dialogService.OpenFile("Seleziona file immagine", "Immagini (*.jpg;*.png)|*.jpg;*.png")
        If Not String.IsNullOrEmpty(_imgProfilo) Then
            Dim img = Image.FromFile(_imgProfilo)
            If img.Width > MAX_IMAGE_WIDTH OrElse img.Height > MAX_IMAGE_HEIGHT Then
                dialogService.ShowNotification($"Le dimensioni della immagine devono esesere di {MAX_IMAGE_WIDTH} x {MAX_IMAGE_HEIGHT} o inferiori.")
                Exit Sub
            End If
            imgProfilo = _imgProfilo
        End If
    End Sub
#End Region

#Region "Comando per aprire immagini"
    Private _opApriImmagine As ICommand
    Public ReadOnly Property OpApriImmagine As ICommand
        Get
            Return If(_opApriImmagine, New RelayCommand(Of MessaggioChat)(Sub(m) ApriImmagine(m)))
        End Get
    End Property

    Private Sub ApriImmagine(ByVal messaggio As MessaggioChat)
        Dim img = messaggio.Immagine
        If (String.IsNullOrEmpty(img) OrElse Not File.Exists(img)) Then Exit Sub
        Process.Start(img)
    End Sub
#End Region

#Region "Gestione Eventi (Event handlers)"
    Private Sub NuovoMsgTesto(nome As String, testo As String, tipoMsg As TipoDiMessaggio)
        If tipoMsg = TipoDiMessaggio.Unicast Then
            Dim msgChat As New MessaggioChat With {.Autore = nome, .Testo = testo, .DataOra = DateTime.Now}
            Dim mittente = _partecipanti.Where(Function(u) String.Equals(u.Nome, nome)).FirstOrDefault
            ctxTaskFactory.StartNew(Sub() mittente.Chatter.Add(msgChat)).Wait()

            If Not (PartecipanteSelezionato IsNot Nothing AndAlso mittente.Nome.Equals(PartecipanteSelezionato.Nome)) Then
                ctxTaskFactory.StartNew(Sub() mittente.HaMandatoNuovoMsg = True).Wait()
            End If
        End If
    End Sub

    Private Sub NuovoMsgImmagine(nome As String, immagine As Byte(), tipoMsg As TipoDiMessaggio)
        If tipoMsg = TipoDiMessaggio.Unicast Then
            Dim CartellaImmagini = Path.Combine(Environment.CurrentDirectory, "Image Messages")
            If Not Directory.Exists(CartellaImmagini) Then Directory.CreateDirectory(CartellaImmagini)

            Dim contImmagini = Directory.EnumerateFiles(CartellaImmagini).Count() + 1
            Dim percorsoImmagine = Path.Combine(CartellaImmagini, $"IMG_{contImmagini}.jpg")

            Dim convertitoreImmagini As New ImageConverter
            Using img As Image = CType(convertitoreImmagini.ConvertFrom(immagine), Image)
                img.Save(percorsoImmagine)
            End Using

            Dim msgChat As New MessaggioChat With {.Autore = nome, .Immagine = percorsoImmagine, .DataOra = DateTime.Now}
            Dim mittente = _partecipanti.Where(Function(u) String.Equals(u.Nome, nome)).FirstOrDefault
            ctxTaskFactory.StartNew(Sub() mittente.Chatter.Add(msgChat)).Wait()

            If Not (PartecipanteSelezionato IsNot Nothing AndAlso mittente.Nome.Equals(PartecipanteSelezionato.Nome)) Then
                ctxTaskFactory.StartNew(Sub() mittente.HaMandatoNuovoMsg = True).Wait()
            End If
        End If
    End Sub

    Private Sub LoginPartecipante(ByVal u As Utente)
        Dim _partecipante = Partecipanti.FirstOrDefault(Function(partecipante) String.Equals(partecipante.Nome, u.Nome))
        If _loggato AndAlso _partecipante Is Nothing Then
            ctxTaskFactory.StartNew(Sub() Partecipanti.Add(New Partecipante With {.Nome = u.Nome, .imgProfilo = u.imgProfilo})).Wait()
        End If
    End Sub

    Private Sub DisconnessionePartecipante(ByVal nome As String)
        Dim tizio = Partecipanti.Where(Function(partecipante) String.Equals(partecipante.Nome, nome)).FirstOrDefault
        If tizio IsNot Nothing Then tizio.Loggato = False
    End Sub

    Private Sub RiconnessionePartecipante(ByVal nome As String)
        Dim tizio = Partecipanti.Where(Function(p) String.Equals(p.Nome, nome)).FirstOrDefault
        If tizio IsNot Nothing Then tizio.Loggato = True
    End Sub

    Private Sub Riconnessione()
        Connesso = False
        Loggato = False
    End Sub

    Private Async Sub Riconnesso()
        Dim imgProfilo = Avatar()
        If Not String.IsNullOrEmpty(_nomeUtente) Then Await chatService.AccessoAsincr(_nomeUtente, imgProfilo)
        Connesso = True
        Loggato = True
    End Sub

    Private Async Sub Disconnesso()
        Dim taskConnessione = chatService.ConnessioneAsync()
        Await taskConnessione.ContinueWith(Sub(task)
                                               If Not task.IsFaulted Then
                                                   Connesso = True
                                                   chatService.AccessoAsincr(_nomeUtente, Avatar()).Wait()
                                                   Loggato = True
                                               End If
                                           End Sub)
    End Sub

    Private Sub PartecipanteStaScrivendo(ByVal nome As String)
        Dim tizio = Partecipanti.Where(Function(p) String.Equals(p.Nome, nome)).FirstOrDefault()
        If tizio IsNot Nothing AndAlso Not tizio.StaScrivendo Then
            tizio.StaScrivendo = True
            Observable.Timer(TimeSpan.FromMilliseconds(1500)).Subscribe(Sub(t) tizio.StaScrivendo = False)
        End If
    End Sub
#End Region

    Private Function Avatar() As Byte()
        Dim imgProfilo As Byte() = Nothing
        If Not String.IsNullOrEmpty(_imgProfilo) Then imgProfilo = File.ReadAllBytes(_imgProfilo)
        Return imgProfilo
    End Function

    Public Sub New(chatSvc As IChatService, diagSvc As IDialogService)
        dialogService = diagSvc
        chatService = chatSvc
        AddHandler chatSvc.NuovoMsgTesto, AddressOf NuovoMsgTesto
        AddHandler chatSvc.NuovoMsgImmagine, AddressOf NuovoMsgImmagine
        AddHandler chatSvc.PartecipanteHaFattoAccesso, AddressOf LoginPartecipante
        AddHandler chatSvc.PartecipanteSiEDisconnesso, AddressOf DisconnessionePartecipante
        AddHandler chatSvc.PartecipanteDisconnesso, AddressOf DisconnessionePartecipante
        AddHandler chatSvc.PartecipanteRiconnesso, AddressOf RiconnessionePartecipante
        AddHandler chatSvc.PartecipanteStaScrivendo, AddressOf PartecipanteStaScrivendo
        AddHandler chatSvc.ConnessioneRiconnessione, AddressOf Riconnessione
        AddHandler chatSvc.ConnessioneRistabilita, AddressOf Riconnesso
        AddHandler chatSvc.ConnessioneChiusa, AddressOf Disconnesso

        ctxTaskFactory = New TaskFactory(TaskScheduler.FromCurrentSynchronizationContext)
    End Sub
End Class